Add introducing reciprocal rank fusion for hybrid search blog (#3621)
opensearch-project/project-website: _posts/2025-02-04-introducing-reciprocal-rank-fusion-hybrid-search.md

---
layout: post
title:  Introducing reciprocal rank fusion for hybrid search
authors:
  - ryanbogan
  - gaievski
  - minalsha
  - kolchfa
date: 2025-02-05
has_science_table: true
categories:
  - technical-posts

> meta_keywords: OpenSearch hybrid search, reciprocal rank fusion, RRF, search ranking, score normalization, Min-max normalization, L2 normalization, ranking algorithms, data normalization

Reviewer: Please update the meta with the following: 

_meta_keywords:_ reciprocal rank fusion, OpenSearch 2.19, neural search plugin, hybrid search, relevance ranking, search normalization, k-nn search, L2 normalization, how reciprocal rank fusion works

_meta_description:_  Learn about reciprocal rank fusion (RRF) using the Neural Search plugin in OpenSearch 2.19. Discover how this new approach to hybrid search merges results from multiple query sources for improved relevance.

### Resistance to outliers

Min-max and L2 normalization are sensitive to outliers, meaning extreme scores can disproportionately impact final rankings. Since RRF aggregates rankings rather than scores, it prevents anomalous values from distorting relevance.

### Consistency in relevance ranking

L2 normalization aligns scores to a common scale but lacks a mechanism to prioritize documents that appear across multiple queries. RRF excels at this by favoring items that rank highly across diverse query methods, ensuring more reliable relevance ranking.

## Practical applications of RRF

RRF is particularly effective in search scenarios where datasets present specific challenges.

### Handling score variability across query methods

> Different search methods—such as BM25, neural search, and k-NN—produce scores on incompatible scales. Techniques like L2 or Min-max normalization attempt to standardize these scores but can lead to suboptimal rankings. RRF sidesteps this issue by emphasizing rank consistency rather than absolute score alignment.

Reviewer 2: ```suggestion
Different search methods—such as BM25, neural search, and k-NN—produce scores on incompatible scales. Techniques like L2 or min-max normalization attempt to standardize these scores but can lead to suboptimal rankings. RRF sidesteps this issue by emphasizing rank consistency rather than absolute score alignment.
```

> **Example:** In multimodal search pipelines, where text-based queries produce a wide range of scores while visual features generate a narrower range, RRF ensures that smaller-scale signals are not overshadowed.

Reviewer 2: ```suggestion
**Example**: In multimodal search pipelines, where text-based queries produce a wide range of scores while visual features generate a narrower range, RRF ensures that smaller-scale signals are not overshadowed.
```

### Enhancing e-commerce search with sparse behavioral signals

E-commerce datasets often contain sparse behavioral signals, such as user clicks or purchases, which can be difficult to incorporate into search rankings. Rank-based aggregation helps highlight relevant products even when engagement data is limited.

> **Example:** When merging behavioral data with metadata, RRF ensures that niche products with high semantic or metadata relevance remain visible, whereas min-max and L2 normalization may struggle with sparse signals.

Reviewer 2: ```suggestion
**Example**: When merging behavioral data with metadata, RRF ensures that niche products with high semantic or metadata relevance remain visible, whereas min-max and L2 normalization may struggle with sparse signals.
```

### Managing noisy or outlier-prone data

Datasets with high variance or frequent outliers---such as scientific research or log data---pose challenges for score-based methods like min-max or L2. RRF prevents outliers from distorting search results by focusing on rank rather than score.

> **Example:** In scientific datasets, metadata from top-tier journals often skews scores. RRF integrates these results without overemphasizing outliers, leading to more balanced rankings.

Reviewer 2: ```suggestion
**Example**: In scientific datasets, metadata from top-tier journals often skews scores. RRF integrates these results without overemphasizing outliers, leading to more balanced rankings.
```

### Supporting dynamic or evolving data

Rapidly changing datasets, such as streaming logs, require frequent recalibration for L2 and min-max normalization, which can introduce instability or latency. RRF maintains stable rankings by aggregating based on static rank positions.

> **Example:** In log search pipelines, RRF consistently prioritizes frequently occurring patterns, even as scoring distributions shift over time.

Reviewer 2: ```suggestion
**Example**: In log search pipelines, RRF consistently prioritizes frequently occurring patterns, even as scoring distributions shift over time.
```

## Comparing RRF with score-based methods

> The following table compares two hybrid search approaches using the same query: a standard hybrid search pipeline and an RRF-based hybrid search pipeline. We selected the top three and bottom three results for comparison. Notice that RRF provides more consistent scores across documents because it ranks them based on relative rank (position) of a document within each query result rather than raw scores.

Reviewer 2: ```suggestion
The following table compares two hybrid search approaches using the same query: a standard hybrid search pipeline and an RRF-based hybrid search pipeline. We selected the top three and bottom three results for comparison. Notice that RRF provides more consistent scores across documents because it ranks them based on the relative rank (position) of a document within each query result rather than on raw scores.
```

## How RRF works

RRF ranks documents by performing the following steps:

1. **Sort documents by score**: Each query method sorts documents by score on every shard.

   ![Initial document scores on each shard](/assets/media/blog-images/2025-02-04-introducing-reciprocal-rank-fusion-hybrid-search/RRFInitialShardScores.png){: .img-fluid}

2. **Assign rank positions**: Documents are ranked based on score for each query.

   ![Document rankings for each query](/assets/media/blog-images/2025-02-04-introducing-reciprocal-rank-fusion-hybrid-search/RRFQueryDocRankings.png){: .img-fluid}

3. **Apply the RRF formula**: The RRF score is computed using the following formula:
   
   ```python
   rankScore(document_i) = sum((1/(k + query_1_rank), (1/(k + query_2_rank), ..., (1/(k + query_j_rank)))
   ```
   
   In this formula, `k` is a rank constant, and `query_j_rank` represents the ranking of a document in a particular query method. The example in the following diagram applies this formula using the default rank constant of 60.

   ![Applying RRF formula](/assets/media/blog-images/2025-02-04-introducing-reciprocal-rank-fusion-hybrid-search/RRFNormalization.png){: .img-fluid}

> 4. **Add rank contributions:** Rank calculations are combined, and documents are sorted by decreasing rank score.

Reviewer 2: ```suggestion
4. **Add rank contributions**: Rank calculations are combined, and documents are sorted by decreasing rank score.
```

   ![Final rank score calculations](/assets/media/blog-images/2025-02-04-introducing-reciprocal-rank-fusion-hybrid-search/RRFFinalScoreCalculations.png){: .img-fluid}

> 5. **Return the top results:** The highest-ranked documents are retrieved based on the query size.

Reviewer 2: ```suggestion
5. **Return the top results**: The highest-ranked documents are retrieved based on the query size.
```

```json
PUT /_search/pipeline/rrf-pipeline
{
  "description": "Post processor for hybrid RRF search",
  "phase_results_processors": [
    {
      "score-ranker-processor": {
        "combination": {
          "technique": "rrf"
        }
      }
    }
  ]
}
```

You can also specify the rank constant as part of the pipeline; the rank constant must be `1` or greater. Larger rank constants make the scores more uniform, reducing the impact of top-ranked items. Smaller rank constants create steeper differences between ranks, giving much more weight to top-ranked items. By default, the rank constant is set to `60`: 

```json
PUT /_search/pipeline/rrf-pipeline
{
  "description": "Post processor for hybrid RRF search",
  "phase_results_processors": [
    {
      "score-ranker-processor": {
        "combination": {
          "technique": "rrf",
          "rank_constant": 40
        }
      }
    }
  ]
}
```

Next, create a hybrid query and apply the pipeline to it:

```json
POST my_index/_search?search_pipeline=rrf-pipeline
{
   "query": {
     "hybrid": [
         {}, // First Query
         {}, // Second Query
         ... // Other Queries
     ] 
   }
}
```

> ## Benchmarking RRF Performance

Reviewer 2: ```suggestion
## Benchmarking RRF performance
```

> Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes. To assess RRF’s performance comprehensively, we measured three key metrics across six distinct datasets. For information about datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).

Reviewer 2: ```suggestion
Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes. To assess RRF's performance comprehensively, we measured three key metrics across six distinct datasets. For information about the datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).
```

Search relevance was quantified using the industry-standard Normalized Discounted Cumulative Gain at rank 10 (NDCG@10). We also tracked system performance using search latency measurements and monitored CPU utilization to analyze resource consumption during the experiments. This setup provided a strong foundation for evaluating both search quality and operational efficiency.

### NDCG@10

> The following table compares the NDCG@10 scores across different search methods (BM25, Neural, Hybrid, and Hybrid with RRF) for various datasets. The percent difference column shows the relative performance change between Hybrid and Hybrid with RRF approaches.

Reviewer 2: ```suggestion
The following table compares the NDCG@10 scores across different search methods (BM25, Neural, Hybrid, and Hybrid with RRF) for various datasets. The *Percent difference* column shows the relative performance change between the Hybrid and Hybrid with RRF approaches.
```

|            | BM25     | Neural  | Hybrid  | Hybrid with RRF | Percent difference |
|------------|----------|---------|---------|-----------------|--------------------|
| NFCorpus   | 	0.3065  | 	0.2174 | 	0.3076 | 	0.2977         | 	3.22%             |
| ArguAna    | 	0.4258  | 	0.4239 | 	0.4507 | 	0.4476         | 	0.69%             |
| FIQA	      | 0.2389	  | 0.2004	 | 0.2693  | 	0.2474         | 	8.13%             |
| Trec-Covid | 	0.6087	 | 0.2718  | 	0.5905 | 	0.5877         | 	0.47%             |
| SciDocs    | 	0.155   | 	0.1075 | 	0.1602 | 	0.1525         | 	4.81%             |
| Quora	     | 0.7424	  | 0.8256	 | 0.8452	 | 0.796	          | 5.82%              |
|            |          |         |         | **Average:**	   | 3.86%              |

### Search latency 

> The following table presents search latency measurements in milliseconds at different percentiles (p50, p90, and p99) for both Hybrid and RRF approaches. The percent difference columns show the relative performance impact between these methods.

Reviewer 2: ```suggestion
The following table presents search latency measurements in milliseconds at different percentiles (p50, p90, and p99) for both the Hybrid and RRF approaches. The *Percent difference* columns show the relative performance impact between these methods.
```

    <td>424</td>
    <td>-0.12%</td>
    <td>459.475</td>
    <td>458.5</td>
    <td>0.21%</td>
  </tr>
  <tr>
    <td>FIQA</td>
    <td>109</td>
    <td>105.25</td>
    <td>3.44%</td>
    <td>139.65</td>
    <td>137.5</td>
    <td>1.54%</td>
    <td>165.76</td>
    <td>160.79</td>
    <td>3.00%</td>
  </tr>
  <tr>
    <td>Trec-Covid</td>
    <td>165.5</td>
    <td>159.75</td>
    <td>3.47%</td>
    <td>209.5</td>
    <td>203.6</td>
    <td>2.82%</td>
    <td>240.73</td>
    <td>237.78</td>
    <td>1.23%</td>
  </tr>
  <tr>
    <td>SciDocs</td>
    <td>103</td>
    <td>103.5</td>
    <td>-0.49%</td>
    <td>126.5</td>
    <td>126.05</td>
    <td>0.36%</td>
    <td>156.02</td>
    <td>159.015</td>
    <td>-1.92%</td>
  </tr>
  <tr>
    <td>Quora</td>
    <td>167</td>
    <td>170</td>
    <td>-1.80%</td>
    <td>209.5</td>
    <td>212.55</td>
    <td>-1.46%</td>
    <td>264.5</td>
    <td>266.51</td>
    <td>-0.76%</td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td><b>Average:</b></td>
    <td>1.62%</td>
    <td></td>
    <td><b>Average:</b></td>
    <td>1.42%</td>
    <td></td>
    <td><b>Average:</b></td>
    <td>0.78%</td>
  </tr>
</table>

### CPU utilization

> The following table shows CPU utilization measurements on the coordinator node, comparing hybrid and RRF approaches. The percent difference column indicates the relative change in CPU usage between the two methods.

Reviewer 2: ```suggestion
The following table shows CPU utilization measurements on the coordinator node, comparing the Hybrid and RRF approaches. The *Percent difference* column indicates the relative change in CPU usage between the two methods.
```

> 

Reviewer 2: I went with italics for the name of the column, but you can choose different formatting if you'd like.

|            | Hybrid  | RRF          | Percent difference |
|------------|---------|--------------|--------------------|
| NFCorpus	  | 0.783%  | 	0.838%      | 	0.055%            |
| ArguAna	   | 0.844%	 | 0.853%       | 	0.008%            |
| FIQA	      | 0.835%	 | 0.851%       | 	0.016%            |
| Trec-Covid | 	1.406% | 	0.979%      | 	-0.427%           |
| SciDocs	   | 0.745%	 | 0.873%       | 	0.128%            |
| Quora	     | 1.054%	 | 1.076%       | 	0.022%            |
|            |         | **Average:** | -0.033%            |

## Conclusions

> Our benchmark experiments highlight the following RRF’s advantages and trade-offs compared to conventional hybrid search approaches:

Reviewer 2: ```suggestion
Our benchmark experiments highlight the following advantages and trade-offs of RRF compared to conventional hybrid search approaches:
```

## What's next?

Our roadmap includes several important enhancements to RRF:

* **Customizable weights**: We plan to implement weight support similar to score-based ranking techniques, allowing more nuanced control over the ranking algorithm. For more information, see [this issue](https://github.com/opensearch-project/neural-search/issues/1152). 

* **Better handling of missing items**: Currently, missing items default to a score of 0.0, but this may not be optimal for all use cases. We're exploring multiple approaches for handling missing items, such as configurable default values, using `max_rank + 1` for missing items, or completely ignoring missing items in calculations. For more information, see [this issue](https://github.com/opensearch-project/neural-search/issues/1153).

We are also expanding OpenSearch's hybrid search capabilities beyond RRF by planning the following improvements to our normalization framework:

> * **Z-score normalization**: Adds the popular z-score normalization technique. For more information, see [this PR](https://github.com/opensearch-project/neural-search/pull/470).

Reviewer 2: ```suggestion
* **Z-score normalization**: Adds the popular z-score normalization technique. For more information, see [this pull request](https://github.com/opensearch-project/neural-search/pull/470).
```

> * **Custom normalization functions**: Enables you to define your own normalization logic and allows fine-tuning of search result rankings. For more information, see [this issue](https://github.com/opensearch-project/neural-search/issues/994)

Reviewer 2: ```suggestion
* **Custom normalization functions**: Enables you to define your own normalization logic and allows fine-tuning of search result rankings. For more information, see [this issue](https://github.com/opensearch-project/neural-search/issues/994).
```
